src/Workspaces/Remote/Core/Services/Options/FileWatcher.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.IO;
+using System.Threading;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.VisualStudio.CodingConventions;
+
+namespace Microsoft.CodeAnalysis.Editor.Options
+{
+    [Export(typeof(IFileWatcher)), Shared]
+    internal sealed class FileWatcher : IFileWatcher
+    {
+        private ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher> _watchers = ImmutableDictionary<(string fileName, string directoryPath), FileSystemWatcher>.Empty;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public FileWatcher()
+        {
+        }
+
+        public event ConventionsFileChangedAsyncEventHandler ConventionFileChanged;
+        public event ContextFileMovedAsyncEventHandler ContextFileMoved;
+
+        public void Dispose()
+        {
+            var watchers = Interlocked.Exchange(ref _watchers, null);
+            if (watchers == null)
+            {
+                return;
+            }
+
+            ConventionFileChanged = null;
+            ContextFileMoved = null;
+        }
+
+        public void StartWatching(string fileName, string directoryPath)
+        {
+            FileSystemWatcher watcher = null;
+            var updated = ImmutableInterlocked.Update(
+                ref _watchers,
+                watchers =>
+                {
+                    if (!watchers.TryGetValue((fileName, directoryPath), out var existingWatcher))
+                    {
+                        if (watcher == null)
+                        {
+                            watcher = new FileSystemWatcher(directoryPath, fileName);
+                        }
+
+                        return watchers.Add((fileName, directoryPath), watcher);
+                    }
+
+                    return watchers;
+                });
+
+            if (updated)
+            {
+                FileSystemEventHandler handler = (sender, e) => HandleFileSystemEvent(fileName, directoryPath, e);
+                watcher.Changed += handler;
```

Comment:
are we saying dispose will unsbuscribe all these event handler?

Reply:
➡️ Yes, it does

Reply:
I guess since FileSystemWatcher is BCL type, we can rely on implementation detail..

Reply:
> I guess since FileSystemWatcher is BCL type, we can rely on implementation detail..

Only chiming in to say that saying that makes the BCL team unhappy. :smile:

Reply:
I will correct my previous statement to not rely on implementation details.

It is irrelevant to this implementation whether or not `FileSystemWatcher` unsubscribes event handlers when `Dispose` is called.